pywinauto/controls/HwndWrapper.py

```diff
@@ -1564,17 +1565,69 @@ def GetFocus(self):
 
     #-----------------------------------------------------------
     def SetFocus(self):
-        """Set the focus to this control
 
-        Bring the window to the foreground first if necessary."""
+        """
+        Set the focus to this control.
+        Bring the window to the foreground first if necessary.
+        """
 
         # find the current foreground window
-        cur_foreground = win32functions.GetForegroundWindow()
+        cur_foreground = win32gui.GetForegroundWindow()
 
         # if it is already foreground then just return
         if self.handle != cur_foreground:
             # set the foreground window
-            win32gui.SetForegroundWindow(self.handle)
+
+            # get the thread of the window that is in the foreground
+            cur_fore_thread = win32process.GetWindowThreadProcessId(
+                cur_foreground)[0]
+
+            # get the thread of the window that we want to be in the foreground
+            control_thread = win32process.GetWindowThreadProcessId(
+                self.handle)[0]
+
+            # if a different thread owns the active window
+            if cur_fore_thread != control_thread:
+                # Attach the two threads and set the foreground window
+                win32process.AttachThreadInput(control_thread,
+                                               cur_fore_thread,
+                                               1)
+
+                self.actions.log('Call SetForegroundWindow within attached '
+                                 'threads - {0} & {1}.'.format(control_thread,
+                                                               cur_fore_thread,
+                                                               ))
+                win32gui.SetForegroundWindow(self.handle)
+
+                # ensure foreground window has changed to the target
+                # or is 0(no foreground window) before the threads detaching
+                timings.WaitUntil(
+                    Timings.setfocus_timeout,
+                    Timings.setfocus_retry,
+                    lambda: win32gui.GetForegroundWindow()
+                        in [self.TopLevelParent().handle, 0])
+
+                # get the threads again to check they are still valid.
+                cur_fore_thread = win32process.GetWindowThreadProcessId(
+                    cur_foreground)[0]
+                control_thread = win32process.GetWindowThreadProcessId(
+                    self.handle)[0]
+
+                if cur_fore_thread and control_thread:  # both are valid
+                    # Detach the threads
+                    win32process.AttachThreadInput(control_thread,
+                                                   cur_fore_thread,
+                                                   0)
+            else:
+                # same threads - just set the foreground window
+                self.actions.log('Call SetForegroundWindow within one thread.')
```

Comment:
ActionLogger is supposed to be a user actions logger, not a debug one. Some warnings can also be logged with ActionLogger. But internal debug info is not a target for that feature.